### PR TITLE
Feature/controller add turris omnia rdkb support

### PR DESCRIPTION
**Pull request by morantr**
Port controller changes from turris omnia support branch (#307).

### CODE REVIEW

**Review by tomereli — APPROVED** [review]

**Review by tomereli — CHANGES_REQUESTED** [review]
First of all, please add a PR description which summarizes the changes

File: controller/CMakeLists.txt
Comment (arnout) [inline controller/CMakeLists.txt:71]: Instead of passing `-Wno-shift-negative-value`, wouldn't it be better to fix the issue? Should be easy enough to fix.

The other exclusions are OK, they're also used in other places.

File: controller/CMakeLists.txt
Comment (arnout) [inline controller/CMakeLists.txt]: I don't think there's any reason to move CMAKE_C_FLAGS_RELEASE after the compiler-specific flags, it should work either way.

However, I do thing CMAKE_CXX_FLAGS should move, because now it doesn't include the -Wno-address-of-packed-member.

File: controller/src/beerocks/cli/beerocks_cli_main.cpp
Comment (arnout) [inline controller/src/beerocks/cli/beerocks_cli_main.cpp:56]: This change (all changes to this file) should be in a separate commit, with a commit message that explains why.

**Comment by arnout** [comment]
Commit d0b9cea797552e3a34cee229f713bf1ae907ad84 was originally authored by Raphaël (commit 6dc89b8ebd810c9893787d633ecfa0b62a1d2adb). You can't claim you authored it, you must preserve Raphaël's authorship. If you make minor changes to it, you should add your own Signed-off-by but keep the original author. If you make major changes, you can take authorship, but then Raphaël's Signed-off-by should still be there.

To fix this, you can either restore the author with `git commit --amend --author='Raphaël Mélotte <raphael.melotte@essensium.com>'`, or you can recover the original metadata with `git commit --amend -s -c 6dc89b8ebd810c9893787d633ecfa0b62a1d2adb`.

**Comment by arnout** [comment]
> you must preserve Raphaël's authorship.

There are others as well, e.g. 26f3b5cebe29ef4c06ace57f2c5932d6a1e03844 was authored by @vitalybu. Please review all commits and restore authorship.

File: controller/src/beerocks/cli/beerocks_cli_bml.cpp
Comment (arnout) [inline controller/src/beerocks/cli/beerocks_cli_bml.cpp]: I believe this change is no longer needed because it has been fixed in the analyzer. @rmelotte can you confirm?

File: controller/src/beerocks/master/tasks/channel_selection_task.cpp
Comment (arnout) [inline controller/src/beerocks/master/tasks/channel_selection_task.cpp]: There's an indentation error introduced here (in commit "controller: Add support for passive mode") that gets fixed in a later commit.

File: controller/src/beerocks/master/tasks/optimal_path_task.cpp
Comment (arnout) [inline controller/src/beerocks/master/tasks/optimal_path_task.cpp:348]: This is a good example of a change that should be done in a completely separate commit, with a good commit message that explains why 50% is "better" than 100%. I guess the reasoning is something like: "Some STAs are able to perform 11k beacon requests, but not for all cases, e.g. not in a different band. In this case, we still want to use the 11k responses we do get."

File: controller/src/beerocks/master/tasks/optimal_path_task.cpp
Comment (arnout) [inline controller/src/beerocks/master/tasks/optimal_path_task.cpp:96]: I haven't tested this, but this looks like optimal path task will be disabled entirely except for the VAP+sta_mac combinationas that are explicitly mentioned in the controller config file. I don't see how that can be a useful feature...

File: controller/src/beerocks/master/beerocks_master_mrouter.cpp
Comment (arnout) [inline controller/src/beerocks/master/beerocks_master_mrouter.cpp:121]: More a coding style question than an actual review comment:

Somebody (@tomereli ?) said at some point in reviews that for smartptrs, you should use `if (!ptr)` rather than `if (ptr == nullptr)`. If that is the case, I think we should do the same for normal pointers.

Whoever made that comment should update CODINGSTYLE.md with that information.

File: controller/src/beerocks/master/db/network_map.cpp
Comment (arnout) [inline controller/src/beerocks/master/db/network_map.cpp]: GET_MESSAGE_POINTER cannot yield nullptr because it adds size to the pointer. Well, it can be nullptr if the pointer happens to be minus maxint, but that is unlikely...

In other words, I don't think this check is needed.

Same in a few other places.

File: controller/src/beerocks/master/db/network_map.cpp
Comment (arnout) [inline controller/src/beerocks/master/db/network_map.cpp:501]: I don't know how relevant this check is, because there is no way that event can be nullptr. It can only happen if init() wasn't called (or the argument to buffer() is larger than the allocated size of buffer, but since it's 0 that can't happen).

But OK, no harm I guess.

**Review by arnout — APPROVED** [review]
As usual, I have a lot of feedback, but nothing critical, so I'm approving. Please do the following:

- rebase on master;
- remove the analyzer speedup and load_steer_on_vaps patches;
- correct authorship.

If the removed patches are needed after all, we can still recover them from the dev/turris-omnia branch.

File: controller/src/beerocks/cli/beerocks_cli_bml.cpp
Comment (rmelotte) [reply controller/src/beerocks/cli/beerocks_cli_bml.cpp]: It's not needed anymore with the "controller: remove "unknown IP" states" commit (currently d0b9cea).
It was only needed because the controller didn't send updates to the analyzer when an event occurred (a client connected, disconnected,...).

File: controller/config/beerocks_controller.conf.in
Comment (tomereli) [inline controller/config/beerocks_controller.conf.in:33]: Can you explain in the commit message what are the diagnostics measurements and why they are enabled now?

File: controller/src/beerocks/bml/bml_defs.h
Comment (tomereli) [inline controller/src/beerocks/bml/bml_defs.h:68]: Wasn't this commit by @rmelotte ?

File: controller/src/beerocks/master/db/db.cpp
Comment (tomereli) [inline controller/src/beerocks/master/db/db.cpp:2205]: The whole of this commit should be removed. The only mode we support from now on is passive mode, so the controller doen't need to be notified.

File: controller/src/beerocks/cli/beerocks_cli_bml.cpp
Comment (tomereli) [reply controller/src/beerocks/cli/beerocks_cli_bml.cpp]: @morantr please remove this commit in that case

File: controller/src/beerocks/master/tasks/optimal_path_task.cpp
Comment (tomereli) [reply controller/src/beerocks/master/tasks/optimal_path_task.cpp:348]: I would go even further and add this explanation here in a comment. Makes reading the code much easier.

File: controller/src/beerocks/master/tasks/optimal_path_task.cpp
Comment (tomereli) [reply controller/src/beerocks/master/tasks/optimal_path_task.cpp:96]: This feature came as a request from a customer (a big one). Perhaps the reasoning is that for some clients / VAP combinations, they might have different algorithms than the optimal path to do the steering?

File: controller/src/beerocks/master/beerocks_master_mrouter.cpp
Comment (tomereli) [reply controller/src/beerocks/master/beerocks_master_mrouter.cpp:121]: @morantr I think you said that KW fails on such checks, and we need to use `if (!ptr)` rather than `if (ptr == nullptr)`.

File: controller/src/beerocks/master/db/network_map.cpp
Comment (tomereli) [reply controller/src/beerocks/master/db/network_map.cpp:501]: And here as well - please change to `if (!event)`

File: controller/src/beerocks/master/tasks/task.cpp
Comment (tomereli) [inline controller/src/beerocks/master/tasks/task.cpp:43]: This looks like a real hack and I'm not sure it's worth it. Why did we do this change in the demo? At least we should add a print when such a thing happens, but I rather drop this commit altogether.

File: controller/src/beerocks/master/db/db.cpp
Comment (tomereli) [inline controller/src/beerocks/master/db/db.cpp:2932]: How often does this happen? We don't want the log to become too clattered.
If you decide to keep this commit, I suggest explaining the scenario in which you felt this print would help so anyone reading this can understand why it is actually needed (me included).

**Review by tomereli — CHANGES_REQUESTED** [review]

File: controller/src/beerocks/cli/beerocks_cli_main.cpp
Comment (morantr) [reply controller/src/beerocks/cli/beerocks_cli_main.cpp:56]: I asked @vitalybu about this, and he said that it didn't compile without it on Turris without it.
That is the reason it is not in a separate commit.

File: controller/src/beerocks/master/tasks/channel_selection_task.cpp
Comment (morantr) [reply controller/src/beerocks/master/tasks/channel_selection_task.cpp]: will do clang-format on that commit

File: controller/src/beerocks/cli/beerocks_cli_bml.cpp
Comment (morantr) [reply controller/src/beerocks/cli/beerocks_cli_bml.cpp]: done

File: controller/src/beerocks/master/tasks/optimal_path_task.cpp
Comment (morantr) [reply controller/src/beerocks/master/tasks/optimal_path_task.cpp:348]: I wish I had a good explanation for it, but there isn't one. This is an optimization that @kantera800 asked me to do since most of the time we didn't get 100%, and optimal_path stopped, so changing to 50% made the demo work better, but it isn't necessarily "better" because the reason we don't have 100% might be because the station "didn't want" to reply. So our decision might be wrong.
So I wouldn't write it in the commit message nor in a comment (stupid thing to write IMO).
But if you have and a specific comment you want me to add as a comment or to the commit message, please write it down and I'll add it.

Either way, all of the changes in this commit are pretty small and it has been done a part of optimal_path optimization so I preferred doing it all in a single commit. I don't think that separating it will give extra benefit, and there is no point in doing it now.

File: controller/src/beerocks/master/tasks/optimal_path_task.cpp
Comment (morantr) [reply controller/src/beerocks/master/tasks/optimal_path_task.cpp:96]: @tomereli is that a question for me? Because I don't have a good answer.

File: controller/src/beerocks/master/beerocks_master_mrouter.cpp
Comment (morantr) [reply controller/src/beerocks/master/beerocks_master_mrouter.cpp:121]: Yes, I said it. But it only applies to smart pointers, because I remember tons of complaints about it when fixing Klocwork issues. For some reason, Klocwork didn't like.
Anyway, it does not apply to regular pointers which is the case here.

File: controller/src/beerocks/master/db/network_map.cpp
Comment (morantr) [reply controller/src/beerocks/master/db/network_map.cpp:501]: But it is not a smart pointer, so it is not relevant.
Now that I look at it, I notice that the cast needs to be changed to a reinterpret_cast.

File: controller/config/beerocks_controller.conf.in
Comment (morantr) [reply controller/config/beerocks_controller.conf.in:33]: I don't know why they enabled it. If I remember correctly you disabled it since it flooded the controller logs with timeout prints.
This task is responsible for periodically ask for statistics on hostapd and it's connected stations.
My guess is that the reason is to see `rx_rssi` value being updated in the connection map.
@vitalybu ?

File: controller/src/beerocks/master/db/db.cpp
Comment (morantr) [reply controller/src/beerocks/master/db/db.cpp:2205]: For now, it is needed because some parts of the code are using it.
I will remove it nicely in a different PR right after I'll finish with the merging the Turris Omnia changes.

File: controller/src/beerocks/master/tasks/task.cpp
Comment (morantr) [reply controller/src/beerocks/master/tasks/task.cpp:43]: It is an optimization for the optimal path task.
Imagine the optimal path task starting to ask a measurement on each BSSID at the time.
Now, let's assume that the task requested measurement on BSSID_A and waited for a response, but reached timeout, and then started to ask for measurement on BSSID_B, but while doing so, received the response for BSSID_A. If it is good, it will be used, and will save an additional measurement to BSSID_A.
If you take into account that each measurement waits for a response to at most 6 seconds and that the second measurement might be bad (when you already received a good measurement), you have a lot to lose here.

File: controller/src/beerocks/master/tasks/optimal_path_task.cpp
Comment (tomereli) [reply controller/src/beerocks/master/tasks/optimal_path_task.cpp:348]: Bottom line - its a HACK we don't know why is working and might not be working at all?
So at least add a comment (and possibly a task) which explains that this is done as part of demo and should be revised in the future.

File: controller/src/beerocks/master/tasks/optimal_path_task.cpp
Comment (tomereli) [reply controller/src/beerocks/master/tasks/optimal_path_task.cpp:96]: It was an answer for @arnout. I think @simonbaren is better equipped to answer that.

File: controller/src/beerocks/master/db/network_map.cpp
Comment (tomereli) [reply controller/src/beerocks/master/db/network_map.cpp:501]: should have been a (nitpick)

File: controller/config/beerocks_controller.conf.in
Comment (tomereli) [reply controller/config/beerocks_controller.conf.in:33]: Please talk to Vitaly and add a comment so we can resolve this conversation.

File: controller/src/beerocks/master/db/db.cpp
Comment (tomereli) [reply controller/src/beerocks/master/db/db.cpp:2205]: OK.

File: controller/src/beerocks/master/tasks/task.cpp
Comment (tomereli) [reply controller/src/beerocks/master/tasks/task.cpp:43]: still feels like a hack, but OK I guess..

**Review by tomereli — APPROVED** [review]
I'm approving so not to block this, since too many development tasks rely on this. Also @arnout already approved so I guess its fine to merge :)

File: controller/src/beerocks/master/db/db.cpp
Comment (morantr) [reply controller/src/beerocks/master/db/db.cpp:2932]: It should happen once when receiving a slave join message which initiates the channel selection.
It also happens when sta is connected, and we update it's channel in the node, but IMO we need to remove the channel from sta node since it will always have the same channel as its parent radio which makes it redundant.

Please try to understand that I didn't write most of the code here, so I don't always have an answer or explanation of why some change has been done, I can only guess.

File: controller/src/beerocks/bml/bml_defs.h
Comment (morantr) [reply controller/src/beerocks/bml/bml_defs.h:68]: re-authored

File: controller/src/beerocks/master/db/network_map.cpp
Comment (morantr) [reply controller/src/beerocks/master/db/network_map.cpp]: removed

File: controller/CMakeLists.txt
Comment (morantr) [reply controller/CMakeLists.txt]: switched between lines 72 and 83

File: controller/CMakeLists.txt
Comment (morantr) [reply controller/CMakeLists.txt:71]: I think @rmelotte did it. @rmelotte do you know how to fix it?

File: controller/config/beerocks_controller.conf.in
Comment (morantr) [reply controller/config/beerocks_controller.conf.in:33]: Ok, @vitalybu confirmed to me that he did it to update the rx_rssi. I'll add it to the commit message.

Comment (morantr) [reply controller/config/beerocks_controller.conf.in:33]: updated the commit message.

File: controller/src/beerocks/master/tasks/optimal_path_task.cpp
Comment (morantr) [reply controller/src/beerocks/master/tasks/optimal_path_task.cpp:348]: added a comment to the code.

**Comment by morantr** [comment]
> Commit [d0b9cea](https://github.com/prplfoundation/prplMesh/commit/d0b9cea797552e3a34cee229f713bf1ae907ad84) was originally authored by Raphaël (commit [6dc89b8](https://github.com/prplfoundation/prplMesh/commit/6dc89b8ebd810c9893787d633ecfa0b62a1d2adb)). You can't claim you authored it, you must preserve Raphaël's authorship. If you make minor changes to it, you should add your own Signed-off-by but keep the original author. If you make major changes, you can take authorship, but then Raphaël's Signed-off-by should still be there.
> 
> To fix this, you can either restore the author with `git commit --amend --author='Raphaël Mélotte <raphael.melotte@essensium.com>'`, or you can recover the original metadata with `git commit --amend -s -c 6dc89b8ebd810c9893787d633ecfa0b62a1d2adb`.

I didn't know I can change authorship on commits I am pushing. Thank you for explaining to me how to do it.
Please try to understand that in order to make this PR I had to use some folder comparing software  (like beyond compare), on the original Turris branch which has 64 commits, and take only the relevant code by moving it from left to right, and then separate nicely to commit which is very hard thing to do, and that I didn't try to claim changes that are not rightfully mine.
I changed the commit authorship on that specific commit you asked me, but not on other commits.
What I wanted to ask is that if some of you see a commit that you think you are the rightful owner of it, please help me and change the authorship.
@vitalybu, @arnout, @rmelotte

**Comment by morantr** [comment]
@rmelotte Do you have an idea why test_flows.sh is failing here? on my machine, it all passes.

**Comment by tomereli** [comment]
> @rmelotte Do you have an idea why test_flows.sh is failing here? on my machine, it all passes.

It's not a real failure since Branch and Pull Request passed. Can we merge this or do we wait for anything else (like commits ownership changes)?

File: controller/CMakeLists.txt
Comment (rmelotte) [reply controller/CMakeLists.txt:71]: I don't remember where the issue was, let me check.

**Comment by rmelotte** [comment]
> It's not a real failure since Branch and Pull Request passed. Can we merge this or do we wait for anything else (like commits ownership changes)?

Is there a timing issue with the CI tests? Can we avoid it or should we just increase the wait time?

File: controller/src/beerocks/master/tasks/optimal_path_task.cpp
Comment (morantr) [reply controller/src/beerocks/master/tasks/optimal_path_task.cpp]: Add a comment in the code

File: controller/src/beerocks/master/beerocks_master_mrouter.cpp
Comment (arnout) [reply controller/src/beerocks/master/beerocks_master_mrouter.cpp:121]: > Anyway, it does not apply to regular pointers which is the case here.

That's why my original comment was: "I think we should do the same for normal pointers." Unless, of course, klockwork doens't like `!ptr` for normal pointers.

**Pull request merged**
